### DISCUSSION TITLE
Listen through container's FX chain in Edit state

### DISCUSSION THREAD

**Discussion by AZpercussion** [MonkeyBars3k/ReaScripts#185] [Ideas]
Very often in sound-design projects I place fx chain on items. Moreoften on glued items.
And all the time it was a pain to propagate fx on multiply items, or store edited group of the items to reglue and place fx on new version.

For now Superglue gives excellent ability to non-destructive put fx on items as on whole one, or make nested fx-ed items.
Also it saves fx chain! Cool! But I can't listen changes through fx while container unglued.

It would be great if script will check is fx chain on the container item take. And if so, it would create children track with the fx chain and place items there.
On glue process it just remove unnecessary children track and move it's fx to container take.

Also it should be for Explode too, I think, cause remove fx chain is easier, than exclude from pool, especially when you work mostly without pooled containers.

**Comment by MonkeyBars3k** [comment]
Interesting idea. I can see the use case. The UX implementation might be quite difficult however, especially considering multichannel use cases. For example, how could this work with media lanes? Unfortunately Reaper still has hard-coded routing limitations (unlike a DAW like Samplitude) that preclude routing items to other tracks freely. 

I guess the restored items could be moved automatically to a child track as you say at the time of Unglue... and of course the superitem fx chain would need to be saved at the start of the Unglue actioin as well, since during the Unglue all the superitem settings are only available as a state chunk.

**Reply by AZpercussion** [reply]
I think, multichannel could work based on container's channel number. Or maybe moved to track fx can increase track channel count.

Saying about fixed lanes... I found several bugs. If they are will fixed, I hope, there will  no problem to implement child track approach.
I'll mail you bugs some later.

**Comment by AZpercussion** [comment]
I tried to make a function that should place before Unglue. The similar will for the after.
There is some troubles with moving FX, but it will be solved somehow. (Solved!)

![Before Unglue](https://user-images.githubusercontent.com/70285853/150789843-6d8d9e18-c85a-489f-b469-ef7fe4cf6407.gif)

`function BeforeUnglue(wantFL)
  local container = reaper.GetSelectedMediaItem(0,0)
  local contTrack = reaper.GetMediaItemTrack(container)
  local contTrackIDX
  local newTrack
  
  for i=0, reaper.CountTracks( 0 ) -1 do
    local testTrack = reaper.GetTrack(0,i)
    if testTrack == contTrack then
      contTrackIDX = i
      break
    end
  end
  
  if contTrackIDX then
    reaper.InsertTrackAtIndex( contTrackIDX+1, false )
    newTrack = reaper.GetTrack(0,contTrackIDX+1)
    --reaper.SetTrackSelected(newTrack, true)
    reaper.SetMediaTrackInfo_Value( newTrack, "I_FOLDERDEPTH", -1 ) --Set child
    reaper.SetMediaTrackInfo_Value( contTrack, "I_FOLDERDEPTH", 1 ) --Set parent
    if wantFL then reaper.SetMediaTrackInfo_Value( newTrack, "I_FREEMODE", wantFL ) end
    --^^Set normal, fixed lanes or FIP mode
    
    ----Now move FX---
    local conTake = reaper.GetActiveTake(container)
    for i=0, reaper.TakeFX_GetCount( conTake ) -1 do
      reaper.TakeFX_CopyToTrack( conTake, 0, newTrack, i, true )
    end
    
    ---Now move container to child track----
    reaper.MoveMediaItemToTrack( container, newTrack )
  end
end

reaper.Undo_BeginBlock2(0)
BeforeUnglue()
reaper.UpdateArrange()
reaper.Undo_EndBlock2(0,"test befor unglue",-1)`

**Reply by AZpercussion** [reply]
Ah! I recognized what I missed. It's just zero instead of index.
reaper.TakeFX_CopyToTrack( conTake, 0, newTrack, i, true )
Updated!

**Comment by MonkeyBars3k** [comment]
This is a terrific idea, thanks for the code! Adding card
